Source/Core/Core/AchievementManager.cpp

```diff
@@ -204,8 +210,82 @@ void AchievementManager::ActivateDeactivateRichPresence()
       nullptr, 0);
 }
 
+void AchievementManager::DoFrame()
+{
+  if (!m_is_game_loaded)
+    return;
+  Core::RunAsCPUThread([&] {
+    rc_runtime_do_frame(
+        &m_runtime,
+        [](const rc_runtime_event_t* runtime_event) {
+          AchievementManager::GetInstance()->AchievementEventHandler(runtime_event);
+        },
+        [](unsigned address, unsigned num_bytes, void* ud) {
+          return AchievementManager::GetInstance()->MemoryPeeker(address, num_bytes, ud);
+        },
+        nullptr, nullptr);
+  });
+  u64 current_time = m_system.GetCoreTiming().GetTicks();
+  if (current_time - m_last_ping_time > SystemTimers::GetTicksPerSecond() * 120)
+    m_queue.EmplaceItem([this] { PingRichPresence(GenerateRichPresence()); });
+}
+
+u32 AchievementManager::MemoryPeeker(u32 address, u32 num_bytes, void* ud)
+{
+  const rc_memory_regions_t* regions = rc_console_memory_regions(m_game_data.console_id);
+  for (u32 ix = 0; ix < regions->num_regions; ix++)
+  {
+    if (address >= regions->region[ix].start_address && address <= regions->region[ix].end_address)
+    {
+      address += (regions->region[ix].real_address - regions->region[ix].start_address);
+      break;
+    }
+  }
+  Core::CPUThreadGuard threadguard(const_cast<Core::System&>(m_system));
+  switch (num_bytes)
+  {
+  case 1:
+    return m_system.GetMMU()
+        .HostTryReadU8(threadguard, address)
+        .value_or(PowerPC::ReadResult<u8>(false, 0u))
+        .value;
+  case 2:
+    return m_system.GetMMU()
+        .HostTryReadU16(threadguard, address)
+        .value_or(PowerPC::ReadResult<u16>(false, 0u))
+        .value;
+  case 4:
+    return m_system.GetMMU()
+        .HostTryReadU32(threadguard, address)
+        .value_or(PowerPC::ReadResult<u32>(false, 0u))
+        .value;
+  case 8:
```

Comment:
[8 is not a valid value for num_bytes to be](https://github.com/RetroAchievements/rcheevos/blob/325e12ddccc271a50883eb49f1a9be8d1370ceff/src/rcheevos/memref.c#L396-L403), so this is unreachable.

Reply:
kind of genuinely impressive that the API doesn't support 64-bit peeks

Reply:
Honestly? It is impressive, and I think I'll keep this in - it wouldn't surprise me if a near-future change allows 8, and this code isn't hurting anything until then.

Reply:
I would doubt that 8 would ever come in the *near-*future, there isn't much reason to actually support 64 bit peeks directly and it can be done indirectly with multiple peeks (which is how RAIntegration does it for their mem search stuff) and the API breakage would be annoying to deal with (especially with emus using RAIntegration with its autoupdate nature).

Regardless, it's probably fine to leave in, maybe note that it's technically unreachable with current rcheevos.